storage/cloudflarestorage.py

```diff
@@ -0,0 +1,97 @@
+import base64
+import logging
+import urllib.parse
+from datetime import datetime, timedelta
+from functools import lru_cache
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization, hashes
+from cryptography.hazmat.primitives.asymmetric import padding
+
+logger = logging.getLogger(__name__)
+
+from storage.cloud import S3Storage
+
+
+class CloudFlareS3Storage(S3Storage):
+    """
+    CloudFlare CDN backed by S3 storage
+    """
+
+    def __init__(
+        self,
+        context,
+        cloudflare_domain,
+        cloudflare_privatekey_filename,
+        storage_path,
+        s3_bucket,
+        *args,
+        **kwargs,
+    ):
+        super(CloudFlareS3Storage, self).__init__(context, storage_path, s3_bucket, *args, **kwargs)
+
+        self.cloudflare_domain = cloudflare_domain
+        self.cloudflare_privatekey = self._load_private_key(cloudflare_privatekey_filename)
+
+    def get_direct_download_url(
+        self, path, request_ip=None, expires_in=60, requires_cors=False, head=False
+    ):
+        # If CloudFront could not be loaded, fall back to normal S3.
+        s3_presigned_url = super(CloudFlareS3Storage, self).get_direct_download_url(
```

Comment:
Put this inside the `if` block, that way it doesn't have to sign _every_ request.

Reply:
Just saw that this is still used even if you're able to sign with CloudFlare keys.
Then the comment `If CloudFront could not be loaded, fall back to normal S3.` doesn't really hold right?

Reply:
This is basically a graceful fallback. We would always set `cloudflare_privatekey`. I could instead throw an error here. I was just following how CloudFront has implemented it

Reply:
Ok. I guess I was confused since you were building the cloudflare_url based off the s3 one...